File: rb/test/account_entity_test.rb
# Account entity test

require "minitest/autorun"
require "json"
require_relative "../Nofrixion_sdk"
require_relative "runner"

class AccountEntityTest < Minitest::Test
  def test_create_instance
    testsdk = NofrixionSDK.test(nil, nil)
    ent = testsdk.Account(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = account_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "update", "load", "remove"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "account." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set NOFRIXION_TEST_ACCOUNT_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    account_ref01_ent = client.Account(nil)
    account_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.account"), "account_ref01"))
    account_ref01_data["account_id"] = setup[:idmap]["account01"]
    account_ref01_data["merchant_id"] = setup[:idmap]["merchant01"]

    account_ref01_data_result = account_ref01_ent.create(account_ref01_data, nil)
    account_ref01_data = Helpers.to_map(account_ref01_data_result)
    assert !account_ref01_data.nil?
    assert !account_ref01_data["id"].nil?

    # LIST
    account_ref01_match = {
      "merchant_id" => setup[:idmap]["merchant01"],
    }

    account_ref01_list_result = account_ref01_ent.list(account_ref01_match, nil)
    assert account_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(account_ref01_list_result),
      { "id" => account_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # UPDATE
    account_ref01_data_up0_up = {
      "id" => account_ref01_data["id"],
    }

    account_ref01_markdef_up0_name = "account_id"
    account_ref01_markdef_up0_value = "Mark01-account_ref01_#{setup[:now]}"
    account_ref01_data_up0_up[account_ref01_markdef_up0_name] = account_ref01_markdef_up0_value

    account_ref01_resdata_up0_result = account_ref01_ent.update(account_ref01_data_up0_up, nil)
    account_ref01_resdata_up0 = Helpers.to_map(account_ref01_resdata_up0_result)
    assert !account_ref01_resdata_up0.nil?
    assert_equal account_ref01_resdata_up0["id"], account_ref01_data_up0_up["id"]
    assert_equal account_ref01_resdata_up0[account_ref01_markdef_up0_name], account_ref01_markdef_up0_value

    # LOAD
    account_ref01_match_dt0 = {
      "id" => account_ref01_data["id"],
    }
    account_ref01_data_dt0_loaded = account_ref01_ent.load(account_ref01_match_dt0, nil)
    account_ref01_data_dt0_load_result = Helpers.to_map(account_ref01_data_dt0_loaded)
    assert !account_ref01_data_dt0_load_result.nil?
    assert_equal account_ref01_data_dt0_load_result["id"], account_ref01_data["id"]

    # REMOVE
    account_ref01_match_rm0 = {
      "id" => account_ref01_data["id"],
    }
    account_ref01_ent.remove(account_ref01_match_rm0, nil)

    # LIST
    account_ref01_match_rt0 = {
      "merchant_id" => setup[:idmap]["merchant01"],
    }

    account_ref01_list_rt0_result = account_ref01_ent.list(account_ref01_match_rt0, nil)
    assert account_ref01_list_rt0_result.is_a?(Array)

    not_found_item = Vs.select(
      Runner.entity_list_to_data(account_ref01_list_rt0_result),
      { "id" => account_ref01_data["id"] })
    assert Vs.isempty(not_found_item)

  end
end

def account_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "account", "AccountTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = NofrixionSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["account01", "account02", "account03", "merchant01", "merchant02", "merchant03", "topup01", "topup02", "topup03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["NOFRIXION_TEST_ACCOUNT_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "NOFRIXION_TEST_ACCOUNT_ENTID" => idmap,
    "NOFRIXION_TEST_LIVE" => "FALSE",
    "NOFRIXION_TEST_EXPLAIN" => "FALSE",
    "NOFRIXION_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["NOFRIXION_TEST_ACCOUNT_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["NOFRIXION_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["NOFRIXION_APIKEY"],
      },
      extra || {},
    ])
    client = NofrixionSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["NOFRIXION_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["NOFRIXION_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/account_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/nofrixion-sdk/go"
	"github.com/voxgig-sdk/nofrixion-sdk/go/core"

	vs "github.com/voxgig-sdk/nofrixion-sdk/go/utility/struct"
)

func TestAccountEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Account(nil)
		if ent == nil {
			t.Fatal("expected non-nil AccountEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := accountBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "account." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set NOFRIXION_TEST_ACCOUNT_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		accountRef01Ent := client.Account(nil)
		accountRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "account"}, setup.data), "account_ref01"))
		accountRef01Data["account_id"] = setup.idmap["account01"]
		accountRef01Data["merchant_id"] = setup.idmap["merchant01"]

		accountRef01DataResult, err := accountRef01Ent.Create(accountRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		accountRef01Data = core.ToMapAny(accountRef01DataResult)
		if accountRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if accountRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		accountRef01Match := map[string]any{
			"merchant_id": setup.idmap["merchant01"],
		}

		accountRef01ListResult, err := accountRef01Ent.List(accountRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		accountRef01List, accountRef01ListOk := accountRef01ListResult.([]any)
		if !accountRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", accountRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(accountRef01List), map[string]any{"id": accountRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		accountRef01DataUp0Up := map[string]any{
			"id": accountRef01Data["id"],
		}

		accountRef01MarkdefUp0Name := "account_id"
		accountRef01MarkdefUp0Value := fmt.Sprintf("Mark01-account_ref01_%d", setup.now)
		accountRef01DataUp0Up[accountRef01MarkdefUp0Name] = accountRef01MarkdefUp0Value

		accountRef01ResdataUp0Result, err := accountRef01Ent.Update(accountRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		accountRef01ResdataUp0 := core.ToMapAny(accountRef01ResdataUp0Result)
		if accountRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if accountRef01ResdataUp0["id"] != accountRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if accountRef01ResdataUp0[accountRef01MarkdefUp0Name] != accountRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", accountRef01MarkdefUp0Name, accountRef01ResdataUp0[accountRef01MarkdefUp0Name])
		}

		// LOAD
		accountRef01MatchDt0 := map[string]any{
			"id": accountRef01Data["id"],
		}
		accountRef01DataDt0Loaded, err := accountRef01Ent.Load(accountRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		accountRef01DataDt0LoadResult := core.ToMapAny(accountRef01DataDt0Loaded)
		if accountRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if accountRef01DataDt0LoadResult["id"] != accountRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		accountRef01MatchRm0 := map[string]any{
			"id": accountRef01Data["id"],
		}
		_, err = accountRef01Ent.Remove(accountRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		accountRef01MatchRt0 := map[string]any{
			"merchant_id": setup.idmap["merchant01"],
		}

		accountRef01ListRt0Result, err := accountRef01Ent.List(accountRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		accountRef01ListRt0, accountRef01ListRt0Ok := accountRef01ListRt0Result.([]any)
		if !accountRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", accountRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(accountRef01ListRt0), map[string]any{"id": accountRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func accountBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "account", "AccountTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read account test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse account test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"account01", "account02", "account03", "merchant01", "merchant02", "merchant03", "topup01", "topup02", "topup03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("NOFRIXION_TEST_ACCOUNT_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"NOFRIXION_TEST_ACCOUNT_ENTID": idmap,
		"NOFRIXION_TEST_LIVE":      "FALSE",
		"NOFRIXION_TEST_EXPLAIN":   "FALSE",
		"NOFRIXION_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["NOFRIXION_TEST_ACCOUNT_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["NOFRIXION_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["NOFRIXION_APIKEY"],
			},
			extra,
		})
		client = sdk.NewNofrixionSDK(core.ToMapAny(mergedOpts))
	}

	live := env["NOFRIXION_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["NOFRIXION_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
